libtenzir/builtins/functions/duration.cpp

```diff
@@ -75,8 +75,181 @@ class plugin final : public function_plugin {
     });
   }
 };
+
+class into_duration_plugin final : public function_plugin {
+public:
+  into_duration_plugin() = default;
+
+  into_duration_plugin(std::string name, tenzir::duration unit)
+    : name_{std::move(name)}, unit_{unit} {
+  }
+
+  auto name() const -> std::string override {
+    return name_;
+  }
+
+  auto make_function(invocation inv, session ctx) const
+    -> failure_or<function_ptr> override {
+    auto expr = ast::expression{};
+    TRY(argument_parser2::function(name())
+          .positional("x", expr, "number")
+          .parse(inv, ctx));
+    return function_use::make([this, expr = std::move(expr)](
+                                evaluator eval, session ctx) -> series {
+      auto b = duration_type::make_arrow_builder(arrow::default_memory_pool());
+      check(b->Reserve(eval.length()));
+      for (const auto& arg : eval(expr)) {
+        match(
+          arg.type,
+          [&](const null_type&) {
+            check(b->AppendNulls(arg.length()));
+          },
+          [&]<class T>(const T&)
+            requires concepts::one_of<T, int64_type, uint64_type, double_type>
+          {
+            constexpr auto min
+              = static_cast<double>(
+                  std::numeric_limits<tenzir::duration::rep>::lowest())
+                - 1.0;
+            constexpr auto max
+              = static_cast<double>(
+                  std::numeric_limits<tenzir::duration::rep>::max())
+                + 1.0;
+            auto overflow = false;
+            for (auto v :
+                 values(T{}, as<type_to_arrow_array_t<T>>(*arg.array))) {
+              if (not v) {
+                check(b->AppendNull());
+                continue;
+              }
+              const auto result
+                = static_cast<double>(v.value()) * unit_.count();
```

Comment:
This leads to a loss of precision for integer types. So anyone working with actual nanosecond counts (=> they probably want precision) will run into issues here when converting them to durations. Not sure whether this is relevant to us right now.

Similarly, one could argue that `count_nanoseconds()` should return `int64_type`. Again, not sure if relevant.

Reply:
That's a very good point and I'm not sure either.

Reply:
Your new version addresses the problem quite nicely, I'd say. Thanks! 

The only remaining thing I can think of (and that could be considered optional because it's only a mild improvement) would be to not `if constexpr` based on `std::same_as<T, std::chrono::nanoseconds>` in `into_duration_plugin`, but instead only based on `not std::same_as<U, double>` (ignoring the unit), and then use `checked_mul` for the multiplication (which will always give you a signed because the unit is signed, so it should also work with `uint64_type` without change). The `double_type` could stay as is. This approach would give you the best precision in all cases, I think. Maybe you can try it if you still have the time.